### ISSUE TITLE
RFE: Preferences stroring need to be consistent

### ISSUE THREAD

**Issue by makiam** [ArtOfIllusion/ArtOfIllusion#54]
Currently I take a look at some plugings and how them saves its configuration.
Base AOI app saves preferences under ${user.home}/.artofillusion folder in .aoiprefs file
SPManager plugin saves its config under ${user.home} in .spmanagerprefs file
HelpPlugin saves preferences under ${user.home}/.artofillusion in HelpPlugin.props file

Both plugins implements very similar to base AOI app code to create save and parse options.
I think we need common API for this 
One more plugin - Polymesh stores options using java preferences api - this is platform specific implementation

**Comment by Sailsman63** [comment]
I agree that this is a bit of a mess. What I'd like to do eventuall is to standardize on the Java preferences API, and then use something like [this file backing store](http://www.davidc.net/programming/java/java-preferences-using-file-backing-store) to keep the options hand-editable.

This would also allow us to create a true portable build that does not need to save anything on the local system - just keep the file in the application directory for the portable build.

**Comment by peteihis** [comment]
One tiny thing is that the things in the aoiprefs don't seem to be in any particular order (and I'm afraid, I'm to blame for some of that). Should take a look at that too.

Out of topic, but there also is a bunch new things, that should be added to preferences. They would need their own tabs ... I'm not sure if those should be written to another file or the same one, with a comment line between... (just for clarity, not for technical reasons).

**Comment by Sailsman63** [comment]
>also is a bunch new things, that should be added to preferences.

What would you like to see added to preferences?

**Comment by peteihis** [comment]
> What would you like to see added to preferences?

I'm sure I had an idea, when I wrote that... :)

But on general level I'd feel more compfortable if parameters that affect different sides of the behaviour would be separated to their dedicated pages. Like the general behaviour (using backups, language....), visual apperance (theme, surface-error ...), view and window default settings (using animations, view default orientations...)

Currently there are not even the user parameters available in the code to handle very many things but there are some that I feel that should be.

One thing I'd like to be added would be a file for last used settings, which should be a separate file from the preferences. There might not be much to put into that, but at least window sizes and positions and the folder paths. To the window size there should be a choice in the preferences to use the last used or a user defined default.

So those are more or less long term ideas again, that need some consideration what goes where and what it takes to happen, but I guess it shoyldnät hurt to keep those in mind.

**Comment by peteihis** [comment]
> @ me: aoiprefs don't seem to be in any particular order

....and there is no way of affecting the order....  it uses `load` and `store` and decides about he orde on it's own.

> @Sailsman63: keep the options hand-editable

Yes please!

>  @Sailsman63: file backing store

I did not spell the code out to myself, but why not... When I read what the spec said about `Properties`, there seemed to be things that are meant to take care or things that pure text format's might have problems with? I suppose that is necessary then... Though hand-editable and special formats don't seem to be the best possible mix? Unfortunately just a plain ASCII text-file is out of the question?

**Comment by Sailsman63** [comment]
> When I read what the spec said about Properties, there seemed to be things that are meant to take care or things that pure text format's might have problems with? I suppose that is necessary then... Though hand-editable and special formats don't seem to be the best possible mix? Unfortunately just a plain ASCII text-file is out of the question?

Not sure what relevance the details of `Properties` spec has here.... the [Preferences API](https://docs.oracle.com/javase/1.5.0/docs/api/java/util/prefs/Preferences.html) handles basically any primitive data type (booleans, floats, etc) by converting them to standard text data. The resulting file should be editable by hand.

The backing store that I referenced earlier does use a `Properties` file, but the `Preferences` API will only feed it text, so no issues.

**Comment by peteihis** [comment]
Well "Pr**r**es" or "Pr***r***es" ... But I was lookin at Preferences of AOI, which are stored as Properties and couple of things came to my mind concernig the hand editing.

I have been wonderig, why the parameters are in a seeminly random order in the .aoiprefs file, though they are nicely grouped in the code. Whan happens is that Properties arranges the keys alphabetically reversed. IMO, this forced automatic reordering is not very hand-editing firendly, especially when the file gets longer. OR then there would have to be a system where every key begins with an index number or somethign of that kind.

Another thing is that this code `p.setProperty("Text Property 1", "Hyvää päivää");` turns into this line in the file: `Text\ Property\ 1=Hyv\u00E4\u00E4 p\u00E4iv\u00E4\u00E4`. Though in preferences you would probably not see other than basic ASCII characters and the back lashes preceding spaces only happen in key texts. Still hand editing that file does not seem the best thing to do. Translation texts are written this way of course but those are not meant for users directly.

Just somethign to think about. :)
